Bottom: c93e2238bbf040aee94664f94184a6137edded12
Top:    69f001f944f5ee8e89bd266c06934b7b7506cb74
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-05-17 17:43:31 +0800

virtio-net: reset num_buf to 1 after linearizing packet

If we successfully linearize the packets, num_buf were set to zero,
but this won't work for the error path which assumes at least one. It
will lead the code try to pop the buffers of next packet and drop
it. Fixing this by set num_buf to 1 if we successfully linearize the
packet.

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/virtio_net.c b/drivers/net/virtio_net.c
index 93ec469..910d2e7 100644
--- a/drivers/net/virtio_net.c
+++ b/drivers/net/virtio_net.c
@@ -536,6 +536,10 @@ static struct page *xdp_linearize_page(struct receive_queue *rq,
 		put_page(p);
 	}
 
+	/* receive_mergeable() assumes there's at least one buffer, so
+	 * align with it.
+	 */
+	*num_buf = 1;
 	/* Headroom does not contribute to packet length */
 	*len = page_off - VIRTIO_XDP_HEADROOM;
 	return page;
